### ISSUE TITLE
Performance degradation when running several optimizations in parallel (SCONE)

### ISSUE THREAD

**Issue by tgeijten** [opensim-org/opensim-core#2959]
There appears to be a performance issue when running several OpenSim 4 optimizations in parallel using SCONE. These are the results of a jumping optimization on a 32 core desktop:

Optimizations | OpenSim3 | OpenSim 4
--- | --- | ---
1 | ~15s | ~16s
3 | ~15s | ~16s
4 | ~16s | ~28s
6 | ~17s | ~30s

Performance appears to degrade consistently above a number of parallel optimizations, when compared to OpenSim 3. Running several optimizations in parallel is a common use case in SCONE.

**Comment by adamkewley** [comment]
@tgeijten , which commit of OpenSim4 is this? Master? And opensim3 refers to your custom SCONE branch?

**Comment by tgeijten** [comment]
@adamkewley yes, sorry for the lack of detail. The optimizations were run based on the master of opensim-core and opensim3-scone. For OpenSim4 I tested both a pre and post #2951 build, but this did not make any difference.

For the test I use a model with only `Thelen2003Muscle`, as to prevent the slowdown caused #2958 (no extra locks are used in SCONE).

I just now ran another set of tests on my 4-core i7 laptop:
Optimizations | OpenSim3 | OpenSim4 | Increase
------------- | -------- | -------- | ---
1 | 25s | 29s | 116%
2 | 43s | 55s | 128%
3 | 64s | 80s | 125%
6 | 130s | 161s | 124%

As you can see, the results are much more consistent and less worrying, even though performance still gets worse with more threads and is worse overal. A benchmark of a single-threaded simulation shows a more modest increase of about 106%. I will perform more tests on the 32 core machine to see if the results are consistent.

@adamkewley did you ever compare OpenSim3 to OpenSim4 with multiple simulations in parallel?

**Comment by adamkewley** [comment]
@tgeijten no :( - most of my perf work has been focused on making OpenSim4 faster. 

I'm *hoping* version 4 is a little (probably not a lot) faster between `4.1` and `master`. It OpenSim shouldn't be slower between versions 3 and 4, but I didn't measure that (v3 was before my time).

**Comment by aymanhab** [comment]
@adamkewley and @tgeijten It is likely the case that OpenSim 4.x is a bit slower than 3.x due to the new flexibility introduced, for example a call to initSystem now constructs a tree of components rather than use a presumed structure, this should be less significant in running simulations and operations post initialization.  Would be great to tease that out using your benchmark framework if possible. (call initSystem repeatedly vs. run long simulation). Not sure how these two scenarios factor in Scone's performance.

**Comment by tgeijten** [comment]
I have performed some additional benchmarks with a longer simulation time (10 secconds). The call to `initSystem()` should be much less significant. Tests ran on a 32 core AMD Ryzen:

Optimizations | OpenSim3 | OpenSim4 | Duration Increase
--- | --- | --- | ---
Single evaluation | 8.3 | 10.4 | 125%
1 | 270 | 365 | 135%
3 | 313 | 461 | 147%
6 | 434 | 699 | 161%

As you can see the slowdown is still large, even though most of the time is spent simulating. See also how the slowdown factor increases when more simulations are run in parallel, which is something I've witnessed throughout the comparison.

This example uses Millard2012EquilibriumMuscle, so #2958 also effects performance for parallel initialization (but this effect be relatively small because of the long simulation time).

Unfortunately, these numbers make it difficult to recommend OpenSim4 over OpenSim3 for SCONE users, unless they really need the new features.

**Comment by aymanhab** [comment]
@adamkewley Having worked on this earlier, do you have a summary of findings. We're planning to investigate this along next version and it would be good to start from where you ended rather than try to reinvent the wheel. Thank you.
